src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioRazorParser.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            // Can be called from any thread
+            var snapshot = _textBuffer.CurrentSnapshot;
+            _parser.QueueChange(null, snapshot);
+        }
+
+        public void Stop()
+        {
+            AssertIsForeground();
+
+            _textBuffer.Changed -= TextBuffer_OnChanged;
+            _parser.Dispose();
+        }
+
+        private void TextBuffer_OnChanged(object sender, TextContentChangedEventArgs contentChange)
+        {
+            AssertIsForeground();
+
+            _idleTimer?.Change(IdleDelay, IdleDelay);
+
+            if (contentChange.Changes.Count > 0)
+            {
+                var firstChange = contentChange.Changes[0];
+                var lastChange = contentChange.Changes[contentChange.Changes.Count - 1];
+
+                var oldLen = lastChange.OldEnd - firstChange.OldPosition;
+                var newLen = lastChange.NewEnd - firstChange.NewPosition;
+
+                var wasChanged = true;
+                if (oldLen == newLen)
+                {
+                    var oldText = contentChange.Before.GetText(firstChange.OldPosition, oldLen);
+                    var newText = contentChange.After.GetText(firstChange.NewPosition, newLen);
+                    wasChanged = !string.Equals(oldText, newText, StringComparison.Ordinal);
+                }
+
+                if (wasChanged)
+                {
+                    var newText = contentChange.After.GetText(firstChange.NewPosition, newLen);
+                    var change = new SourceChange(firstChange.OldPosition, oldLen, newText);
+                    var snapshot = contentChange.After;
+                    var result = PartialParseResultInternal.Rejected;
+
+                    using (_parser.SynchronizeMainThreadState())
+                    {
+                        // Check if we can partial-parse
+                        if (_partialParser != null && _parser.IsIdle)
+                        {
+                            result = _partialParser.Parse(change);
+                        }
+                    }
+
+                    // If partial parsing failed or there were outstanding parser tasks, start a full reparse
+                    if ((result & PartialParseResultInternal.Rejected) == PartialParseResultInternal.Rejected)
+                    {
+                        _parser.QueueChange(change, snapshot);
+                    }
+
+                    if ((result & PartialParseResultInternal.Provisional) == PartialParseResultInternal.Provisional && _idleTimer == null)
+                    {
+                        _idleTimer = new Timer(OnIdle, null, IdleDelay, IdleDelay);
+                    }
+                }
+            }
+        }
+
+        private void OnIdle(object state)
+        {
+            AssertIsBackground();
+
+            if (!_completionBroker.CompletionActive)
+            {
+                _idleTimer.Dispose();
+                _idleTimer = null;
+                Reparse();
```

Comment:
This is pretty wacky. I think we need to keep the timer around rather than creating it every time we do a provisional parse